File: packages/sdk-ts/src/session.ts
import { CookieNexusClient } from './client.js';

/**
 * Session Manager to inject CookieNexus cookies into Node.js HTTP clients and browser automation contexts.
 */
export class SessionManager {
  constructor(private client: CookieNexusClient) {}

  /**
   * Generates or injects Cookie header into a headers map/object.
   */
  async injectHeaders(headers: Record<string, string>, domain: string, vaultIdOverride?: string): Promise<Record<string, string>> {
    const headerStr = await this.client.getCookieHeader(domain, vaultIdOverride);
    const updated = { ...headers };
    if (headerStr) {
      updated['Cookie'] = headerStr;
    }
    return updated;
  }

  /**
   * Injects cookies into an Axios instance default headers.
   */
  async injectAxios(axiosInstance: any, domain: string, vaultIdOverride?: string): Promise<void> {
    const headerStr = await this.client.getCookieHeader(domain, vaultIdOverride);
    if (axiosInstance?.defaults?.headers) {
      if (typeof axiosInstance.defaults.headers.set === 'function') {
        axiosInstance.defaults.headers.set('Cookie', headerStr);
      } else {
        axiosInstance.defaults.headers.common = axiosInstance.defaults.headers.common || {};
        axiosInstance.defaults.headers.common['Cookie'] = headerStr;
      }
    }
  }

  /**
   * Injects Cookie header into a Fetch RequestInit options object.
   */
  async injectFetchOptions(init: any = {}, domain: string, vaultIdOverride?: string): Promise<any> {
    const headerStr = await this.client.getCookieHeader(domain, vaultIdOverride);
    const headers = { ...(init.headers || {}) };
    headers['Cookie'] = headerStr;
    return { ...init, headers };
  }

  /**
   * Injects cookies into a Playwright BrowserContext or Page.
   */
  async injectPlaywright(contextOrPage: any, domain?: string, vaultIdOverride?: string): Promise<void> {
    const pwState = await this.client.getPlaywrightStorageState(domain, vaultIdOverride);
    const target = contextOrPage?.context ? contextOrPage.context() : contextOrPage;
    if (target && typeof target.addCookies === 'function') {
      await target.addCookies(pwState.cookies);
    }
  }

  /**
   * Injects cookies into a Puppeteer Page.
   */
  async injectPuppeteer(page: any, domain?: string, vaultIdOverride?: string): Promise<void> {
    const cookies = await this.client.getCookies(domain, vaultIdOverride);
    if (page && typeof page.setCookie === 'function') {
      const puppeteerCookies = cookies.map((c: any) => ({
        name: c.name,
        value: c.value,
        domain: c.domain ? c.domain.replace(/^\./, '') : undefined,
        path: c.path || '/',
        expires: c.expirationDate || undefined,
        httpOnly: !!c.httpOnly,
        secure: !!c.secure,
        sameSite: c.sameSite || 'Lax',
      }));
      await page.setCookie(...puppeteerCookies);
    }
  }
}

File: packages/sdk-python/cookienexus/session.py
from typing import Optional, Dict, Any
from .client import CookieNexusClient

class SessionManager:
    """
    Session Manager to inject CookieNexus cookies into HTTP client sessions.
    """
    def __init__(self, client: CookieNexusClient):
        self.client = client

    def inject_httpx(self, httpx_client: Any, domain: str, vault_id: Optional[str] = None) -> None:
        """
        Injects cookies directly into an httpx.Client or httpx.AsyncClient.
        """
        cookies = self.client.get_cookies(domain, vault_id=vault_id)
        for c in cookies:
            httpx_client.cookies.set(c['name'], c['value'], domain=c.get('domain'), path=c.get('path', '/'))

    def inject_requests(self, requests_session: Any, domain: str, vault_id: Optional[str] = None) -> None:
        """
        Injects cookies directly into a requests.Session.
        """
        cookies = self.client.get_cookies(domain, vault_id=vault_id)
        for c in cookies:
            requests_session.cookies.set(c['name'], c['value'], domain=c.get('domain'), path=c.get('path', '/'))
